src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java

```diff
@@ -0,0 +1,255 @@
+        if (heads == null || heads.size() == 0) {
+            return false;
+        }
+        Collection<FlowNode> fastEndNodes = convertToFastCheckable(blackList);
+        HashSet<FlowNode> filteredHeads = new HashSet<FlowNode>(heads);
+        filteredHeads.removeAll(fastEndNodes);
+
+        if (filteredHeads.size() == 0) {
+            return false;
+        }
+
+        reset();
+        _blackList = fastEndNodes;
+        setHeads(filteredHeads);
+        return true;
+    }
+
+    /**
+     * Set up for iteration/analysis on a graph of nodes, initializing the internal state
+     * @param head The head FlowNode to start walking back from
+     * @param blackList Nodes that we cannot visit or walk past (useful to limit scanning to only nodes after a specific point)
+     *                  null or empty collection means none
+     * @return True if we can have nodes to work with, otherwise false
+     */
+    public boolean setup(@CheckForNull FlowNode head, @CheckForNull Collection<FlowNode> blackList) {
+        if (head == null) {
+            return false;
+        }
+        return setup(Collections.singleton(head), blackList);
+    }
+
+    public boolean setup(@CheckForNull FlowNode head) {
+        if (head == null) {
+            return false;
+        }
+        return setup(Collections.singleton(head), Collections.EMPTY_SET);
+    }
+
+    /** Public APIs need to invoke this before searches */
+    protected abstract void reset();
+
+    /** Add current head nodes to current processing set, after filtering by blackList */
+    protected abstract void setHeads(@Nonnull Collection<FlowNode> filteredHeads);
+
+    /**
+     * Actual meat of the iteration, get the next node to visit, using & updating state as needed
+     * @param current Current node to use in generating next value
+     * @param blackList Nodes that are not eligible for visiting
+     * @return Next node to visit, or null if we've exhausted the node list
+     */
+    @CheckForNull
+    protected abstract FlowNode next(@Nonnull FlowNode current, @Nonnull Collection<FlowNode> blackList);
+
+    @Override
+    public boolean hasNext() {
+        return _next != null;
+    }
+
+    @Override
+    public FlowNode next() {
+        if (_next == null) {
+            throw new NoSuchElementException();
+        }
+
+        // For computing timings and changes, it may be helpful to keep the previous result
+        // by creating a variable _last and storing _current to it.
+
+//            System.out.println("Current iterator val: " + ((_current == null) ? "null" : _current.getId()));
+//            System.out.println("Next iterator val: " + ((_next == null) ? "null" : _next.getId()));
+        _current = _next;
+        _next = next(_current, _blackList);
+//            System.out.println("New next val: " + ((_next == null) ? "null" : _next.getId()));
+        return _current;
+    }
+
+    @Override
+    public void remove() {
```

Comment:
Should be `final`.

Basically, unless you have a documented and tested use case for overriding a method or class, it should be `final`. (Similarly, methods or types should only be a `public` if you have a documented and tested use case for calling them outside the package.)